File: source/daemon/crates/wardnetd/src/api/devices.rs
use axum::Json;
use axum::extract::{Path, State};
use uuid::Uuid;
use wardnet_types::api::{
    DeviceDetailResponse, DeviceMeResponse, ListDevicesResponse, SetMyRuleRequest,
    SetMyRuleResponse, UpdateDeviceRequest,
};

use crate::api::middleware::{AdminAuth, ClientIp};
use crate::error::AppError;
use crate::state::AppState;

/// GET /api/devices/me
///
/// Thin handler — identifies the caller by source IP and returns their
/// device info and current routing rule. No authentication required.
pub async fn get_me(
    State(state): State<AppState>,
    ClientIp(ip): ClientIp,
) -> Result<Json<DeviceMeResponse>, AppError> {
    let response = state
        .device_service()
        .get_device_for_ip(&ip.to_string())
        .await?;
    Ok(Json(response))
}

/// PUT /api/devices/me/rule
///
/// Thin handler — allows the caller to set their own routing rule.
/// Delegates admin-lock checks to [`DeviceService`](crate::service::DeviceService).
/// No authentication required (self-service by IP).
pub async fn set_my_rule(
    State(state): State<AppState>,
    ClientIp(ip): ClientIp,
    Json(body): Json<SetMyRuleRequest>,
) -> Result<Json<SetMyRuleResponse>, AppError> {
    let response = state
        .device_service()
        .set_rule_for_ip(&ip.to_string(), body.target)
        .await?;
    Ok(Json(response))
}

/// GET /api/devices — List all devices (admin only).
pub async fn list_devices(
    State(state): State<AppState>,
    _auth: AdminAuth,
) -> Result<Json<ListDevicesResponse>, AppError> {
    let devices = state.discovery_service().get_all_devices().await?;
    Ok(Json(ListDevicesResponse { devices }))
}

/// GET /api/devices/:id — Get device detail with routing rule (admin only).
pub async fn get_device(
    State(state): State<AppState>,
    _auth: AdminAuth,
    Path(id): Path<String>,
) -> Result<Json<DeviceDetailResponse>, AppError> {
    let uuid: Uuid = id
        .parse()
        .map_err(|_| AppError::BadRequest("invalid device ID".to_owned()))?;
    let device = state.discovery_service().get_device_by_id(uuid).await?;
    let rule = state
        .device_service()
        .get_device_for_ip(&device.last_ip)
        .await
        .ok()
        .and_then(|r| r.current_rule);
    Ok(Json(DeviceDetailResponse {
        device,
        current_rule: rule,
    }))
}

/// PUT /api/devices/:id — Update device properties (admin only).
///
/// Supports updating name, device type, routing target, and admin-locked flag.
/// Each field is optional; only provided fields are changed.
pub async fn update_device(
    State(state): State<AppState>,
    _auth: AdminAuth,
    Path(id): Path<String>,
    Json(body): Json<UpdateDeviceRequest>,
) -> Result<Json<DeviceDetailResponse>, AppError> {
    let uuid: Uuid = id
        .parse()
        .map_err(|_| AppError::BadRequest("invalid device ID".to_owned()))?;

    // Update name and type if provided.
    let device = state
        .discovery_service()
        .update_device(uuid, body.name.as_deref(), body.device_type)
        .await?;

    let device_id_str = device.id.to_string();

    // Update admin_locked if provided.
    if let Some(locked) = body.admin_locked {
        state
            .device_service()
            .update_admin_locked(&device_id_str, locked)
            .await?;
    }

    // Update routing rule if provided.
    if let Some(target) = body.routing_target {
        state
            .device_service()
            .set_rule(&device_id_str, target)
            .await?;
    }

    // Fetch current rule for the response.
    let rule = state
        .device_service()
        .get_device_for_ip(&device.last_ip)
        .await
        .ok()
        .and_then(|r| r.current_rule);

    Ok(Json(DeviceDetailResponse {
        device,
        current_rule: rule,
    }))
}

File: source/sdk/wardnet-js/src/services/devices.ts
import type { WardnetClient } from "../client.js";
import { apiClient, type ApiClient } from "../internal/client.js";
import type { RoutingTarget } from "../types/device.js";
import type {
  DeviceDetailResponse,
  DeviceMeResponse,
  DeviceProbeResponse,
  DnsCaptureSettingsRequest,
  DnsCaptureSettingsResponse,
  ListDevicesResponse,
  SetMyRuleResponse,
  UpdateDeviceRequest,
} from "../types/api.js";

/** Device management service for the Wardnet daemon. */
export class DeviceService {
  private readonly api: ApiClient;

  constructor(client: WardnetClient) {
    this.api = apiClient(client);
  }

  /** List all devices (admin only). */
  async list(): Promise<ListDevicesResponse> {
    return this.api.get("/devices");
  }

  /** Get a device by ID with its current routing rule (admin only). */
  async getById(id: string): Promise<DeviceDetailResponse> {
    return this.api.get("/devices/{id}", { path: { id } });
  }

  /** Get the calling device's info based on source IP (no auth required). */
  async getMe(): Promise<DeviceMeResponse> {
    const res = await this.api.get("/devices/me");
    // The daemon projects `device` as a bare Device here — without the
    // `dhcp_status` / `current_rule` fields the admin-facing endpoints carry.
    // The public `DeviceMeResponse` has always reused the fuller `Device`
    // type, so narrow the cast to just this field and keep the rest checked.
    return { ...res, device: res.device as DeviceMeResponse["device"] };
  }

  /** Set the calling device's routing rule (no auth required, blocked if admin-locked). */
  async setMyRule(target: RoutingTarget): Promise<SetMyRuleResponse> {
    return this.api.put("/devices/me/rule", { body: { target } });
  }

  /** Update a device's name and/or type (admin only). */
  async update(id: string, body: UpdateDeviceRequest): Promise<DeviceDetailResponse> {
    return this.api.put("/devices/{id}", { path: { id }, body });
  }

  /**
   * Probe a device on the vendor catalog's TCP ports and record whichever
   * answered as identification signals, naming the device if one resolves to a
   * vendor (admin only, issue #1116).
   *
   * This is the only identification signal that sends unsolicited traffic to a
   * device, so per ADR 0025 §5 it must stay bound to an explicit admin action —
   * never a background sweep over `list()`.
   *
   * Rejects with a 409 when the device is not currently on the network: its
   * last known address may since have been handed to someone else, and the
   * resulting vendor would be recorded against the wrong device permanently.
   */
  async identify(id: string): Promise<DeviceProbeResponse> {
    return this.api.post("/devices/{id}/identify", { path: { id } });
  }

  /**
   * Stop managing a device — revert every admin-set configuration to default
   * and return it to unmanaged (admin only, issue #1181).
   *
   * **Destructive.** This revokes the device's Private-DNS grant and its
   * Remote peer credential, disconnecting it. It also clears the device's
   * name, admin lock, DNS capture, routing rule and profiles, DNS-filter
   * settings, DHCP reservation, and zone exceptions, and returns it to the
   * default-for-new zone.
   *
   * Once unmanaged the device becomes subject to device retention and is
   * deleted after 30 days away. Idempotent: a retry after a partial failure
   * completes, and a failure part-way leaves the device still managed rather
   * than half-released.
   */
  async release(id: string): Promise<DeviceDetailResponse> {
    return this.api.post("/devices/{id}/release", { path: { id } });
  }

  /**
   * Assign or clear the household user a device belongs to (admin only,
   * ADR-0031 §4). Pass `null` to clear.
   *
   * **Attribution, never authentication.** The owner's role has no effect on
   * what the device may do — a device caller resolves to the `Device`
   * principal whoever owns it, including an admin. Device identity is derived
   * from the source IP, so treating ownership as a credential would collapse
   * admin access to IP spoofing.
   */
  async setOwner(id: string, ownerUserId: string | null): Promise<DeviceDetailResponse> {
    return this.api.put("/devices/{id}/owner", {
      path: { id },
      body: { owner_user_id: ownerUserId },
    });
  }

  /** Get DNS capture settings and storage stats for a device (admin only). */
  async getDnsCaptureSettings(id: string): Promise<DnsCaptureSettingsResponse> {
    return this.api.get("/devices/{id}/dns-capture", { path: { id } });
  }

  /** Update DNS capture settings for a device (admin only). */
  async updateDnsCaptureSettings(
    id: string,
    body: DnsCaptureSettingsRequest,
  ): Promise<DnsCaptureSettingsResponse> {
    return this.api.patch("/devices/{id}/dns-capture", { path: { id }, body });
  }

  /**
   * Enable or disable DNS capture for the calling device (no auth required,
   * resolved by source IP). Only the `enabled` flag changes — retention caps
   * are admin-only. Returns the device's current capture settings and stats.
   */
  async setMyCaptureEnabled(enabled: boolean): Promise<DnsCaptureSettingsResponse> {
    return this.api.patch("/devices/me/dns-capture", { body: { enabled } });
  }

  /** Acknowledge receipt of DNS events up to and including `upToId`. */
  async ackDnsEvents(upToId: number): Promise<void> {
    await this.api.post("/devices/me/dns-events/ack", { body: { up_to_id: upToId } });
  }
}
